### PR TITLE
[Refactor] reduce sending audit stats in exchange

### LINKED ISSUE #37082: audit stats' lock takes too much time
## Enhancement
starrocks::RuntimeState::intermediate_query_statistic takes about 27% time.

### PR DESCRIPTION
**Pull request by fzhedu**
Why I'm doing:

sending audit stats for each chunk request takes 27% of time in high concurrent workloads.

What I'm doing:

reduce sending audit stats every `_sent_audit_stats_frequency`.  
_sent_audit_stats_frequency = max(64, RoundUpToPowerOfTwo(fragment_ctx->num_drivers() * 4))), it is scale up with drivers' num. It help to know large query' audit stats even missing the last chunk.

the last chunk of exchange should sent audit stats.

results:
reduce 4% cost to null

Fixes https://github.com/StarRocks/starrocks/issues/37082

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

### CODE REVIEW

File: be/src/exec/pipeline/exchange/sink_buffer.cpp [hunk 1]
```diff
@@ -335,10 +309,15 @@ Status SinkBuffer::_try_to_send_rpc(const TUniqueId& instance_id, const std::fun
                     return Status::OK();
                 }
                 // this is the last eos query, set query stats
-                _eos_query_stats[instance_id.lo]->merge_pb(request.params->query_statistics());
-                request.params->clear_query_statistics();
-                _eos_query_stats[instance_id.lo]->to_pb(request.params->mutable_query_statistics());
-                _eos_query_stats[instance_id.lo]->clear();
+                if (auto final_stats = _fragment_ctx->runtime_state()->query_ctx()->intermediate_query_statistic()) {
```
Comment (liuyehcf) [inline be/src/exec/pipeline/exchange/sink_buffer.cpp:312]: what if there is a short circuit, the receiver side won't wait the last eos packet.

File: be/src/exec/pipeline/exchange/sink_buffer.cpp [hunk 2]
```diff
@@ -335,10 +309,15 @@ Status SinkBuffer::_try_to_send_rpc(const TUniqueId& instance_id, const std::fun
...
+                    final_stats->to_pb(request.params->mutable_query_statistics());
+                }
+            }
+        } else {
+            if ((_request_sent & _sent_audit_stats_frequency) == _sent_audit_stats_frequency) {
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/exchange/sink_buffer.cpp:317]: What if total `_request_sent` is less than 127 ?

**Review by murphyatwork — DISMISSED** [review]

File: be/src/exec/pipeline/exchange/sink_buffer.cpp (hunk 2 above)
Comment (fzhedu) [reply be/src/exec/pipeline/exchange/sink_buffer.cpp:317]: that will be up the eos's last request.

File: be/src/exec/pipeline/exchange/sink_buffer.cpp (hunk 1 above)
Comment (fzhedu) [reply be/src/exec/pipeline/exchange/sink_buffer.cpp:312]: it is up to the sent stats by `_sent_audit_stats_frequency `

File: be/src/exec/pipeline/exchange/sink_buffer.h [hunk 3]
```diff
@@ -192,6 +189,8 @@ class SinkBuffer {
     int64_t _first_send_time = -1;
     int64_t _last_receive_time = -1;
     int64_t _rpc_http_min_size = 0;
+
+    int64_t _sent_audit_stats_frequency = 127;
```
Comment (silverbullet233) [inline be/src/exec/pipeline/exchange/sink_buffer.h:193]: better to define a static const variable for 127?

**Review by silverbullet233 — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by fzhedu** [comment]
@mergify backport branch-2.5

**Comment by fzhedu** [comment]
@mergify backport branch-3.0

**Comment by fzhedu** [comment]
@mergify backport branch-3.1

**Comment by fzhedu** [comment]
@mergify backport branch-3.2

**Pull request merged**